runtime/src/chpl-qsbr.c

```diff
@@ -0,0 +1,481 @@
+
+static inline struct tls_node *get_tls_list(void);
+static inline struct tls_node *get_tls_list(void) {
+  return (struct tls_node *) atomic_load_uintptr_t((uintptr_t *) &chpl_qsbr_tls_list);
+}
+
+static inline uint64_t get_epoch(struct tls_node *node);
+static inline uint64_t get_epoch(struct tls_node *node) {
+  return atomic_load_uint_least64_t(&node->epoch);
+}
+
+static inline void set_epoch(struct tls_node *node, uint64_t epoch);
+static inline void set_epoch(struct tls_node *node, uint64_t epoch) {
+  return atomic_store_uint_least64_t(&node->epoch, epoch);
+}
+
+static inline uint64_t get_global_epoch(void);
+static inline uint64_t get_global_epoch(void) {
+  return atomic_load_uint_least64_t(&global_epoch);
+}
+
+static inline uint64_t advance_global_epoch(void);
+static inline uint64_t advance_global_epoch(void) {
+  return atomic_fetch_add_uint_least64_t(&global_epoch, 1);
+}
+
+static inline void observe_epoch(struct tls_node *node);
+static inline void observe_epoch(struct tls_node *node) {
+  atomic_store_uint_least64_t(&node->epoch, get_global_epoch());
+}
+
+static inline uint64_t get_tasks(struct tls_node *node);
+static inline uint64_t get_tasks(struct tls_node *node) {
+  return atomic_load_uint_least64_t(&node->nTasks);
+}
+
+static inline uint64_t add_task(struct tls_node *node);
+static inline uint64_t add_task(struct tls_node *node) {
+  return atomic_fetch_add_uint_least64_t(&node->nTasks, 1);
+}
+
+static inline uint64_t remove_task(struct tls_node *node);
+static inline uint64_t remove_task(struct tls_node *node) {
+  return atomic_fetch_sub_uint_least64_t(&node->nTasks, 1);
+}
+
+static inline void acquire_spinlock(struct tls_node *node);
+static inline void acquire_spinlock(struct tls_node *node) {
+  // Fast path: Quick acquire spinlock
+  if (atomic_exchange_uint_least64_t(&node->spinlock, 1) == 1) {
+    // Slow path: TestAndTestAndSet. Note that we cannot yield or
+    // we will end up calling the checkpoint.
+    while (true) {
+      uint64_t spinlock = atomic_load_uint_least64_t(&node->spinlock);
+      if (spinlock == 0 
+        && atomic_compare_exchange_weak_uint_least64_t(&node->spinlock, 0, 1)) {
+        break;
+      }
+    }
+  }
+}
+
+static inline void release_spinlock(struct tls_node *node);
+static inline void release_spinlock(struct tls_node *node) {
+  atomic_store_uint_least64_t(&node->spinlock, 0);
+}
+
+// Requires spinlock.
+static inline struct defer_node *get_defer_list(struct tls_node *node);
+static inline struct defer_node *get_defer_list(struct tls_node *node) {
+  return node->deferList;
+}
+
+// Requires spinlock. Pops all items less than or equal to 'epoch'.
+static inline struct defer_node *pop_bulk_defer_list(struct tls_node *node, uint64_t epoch);
+static inline struct defer_node *pop_bulk_defer_list(struct tls_node *node, uint64_t epoch) {
+  struct defer_node *dnode = node->deferList, *prev = NULL;
+  while (dnode != NULL) {
+    if (epoch >= dnode->targetEpoch) {
```

Comment:
wouldn't this pop elements equal to epoch? the comment above seems to indicate it wouldn't.

Reply:
The comment says "pops all items *less than or equal* to 'epoch'.", but the more I think about it the more wrong it is. I realize now that it can't be equal to because it receives the minimum epoch of all threads (which means it is still a safe epoch), nice catch there.

Reply:
I'm not sure how much is changed since I last looked, but the comment says pops items equal to epoch, but the code seems to only pop < epoch.

Reply:
Never mind, it looks right, I see it now.